server/src/main/java/keywhiz/service/resources/automation/v2/ClientResource.java

```
@@ -0,0 +1,269 @@
+    });
+
+    // Creates new client record
+    long clientId = clientDAO.createClient(client, creator, request.description());
+
+    // Enrolls client in any requested groups
+    groupsToGroupIds(request.groups())
+        .forEach((maybeGroupId) -> maybeGroupId.ifPresent(
+            (groupId) -> aclDAO.findAndEnrollClient(clientId, groupId)));
+
+    URI uri = UriBuilder.fromResource(ClientResource.class).path(client).build();
+    return Response.created(uri).build();
+  }
+
+  /**
+   * Retrieve listing of client names
+   *
+   * @excludeParams automationClient
+   * @responseMessage 200 List of client names
+   */
+  @GET
+  @Produces(APPLICATION_JSON)
+  public Iterable<String> clientListing(@Auth AutomationClient automationClient) {
+    return clientDAO.getClients().stream()
+        .map(Client::getName)
+        .collect(toSet());
+  }
+
+  /**
+   * Retrieve information on a client
+   *
+   * @excludeParams automationClient
+   * @param name Client name
+   *
+   * @responseMessage 200 Client information retrieved
+   * @responseMessage 404 Client not found
+   */
+  @GET
+  @Path("{name}")
+  @Produces(APPLICATION_JSON)
+  public ClientDetailResponseV2 clientInfo(@Auth AutomationClient automationClient,
+      @PathParam("name") String name) {
+    Client client = clientDAO.getClient(name)
+        .orElseThrow(NotFoundException::new);
+
+    return ClientDetailResponseV2.fromClient(client);
+  }
+
+  /**
+   * Listing of groups accessible to a client
+   *
+   * @excludeParams automationClient
+   * @param name Client name
+   * @return Listing of groups the client has membership to
+   *
+   * @responseMessage 200 Listing succeeded
+   * @responseMessage 404 Client not found
+   */
+  @GET
+  @Path("{name}/groups")
+  @Produces(APPLICATION_JSON)
+  public Iterable<String> clientGroupsListing(@Auth AutomationClient automationClient,
+      @PathParam("name") String name) {
+    Client client = clientDAO.getClient(name)
```

I wouldn't want to throw a web exception in the DAO, so it would mean catching and re-throwing a different exception. For a method that can return something or nothing, Optional<Something> seems like a fit.